### PR TITLE
cpu/cc2538: ADC driver

### PR DESCRIPTION
**Pull request by alignan**
This PR supersedes https://github.com/RIOT-OS/RIOT/pull/5545 and implements the ADC driver for CC2538-based platforms.
- Removed the decimation offsets to return the result in mV
- Added ADC support for the `RE-Mote`, `openmote-cc2538` and `cc2538dk`
- Added SAUL support for the `RE-Mote` and the ADC module, tested with `tests/saul`
- Added `ADC_LINE` macro to iterate ADC channels using `ADC_NUMOF` (required for the `tests/periph_adc`)

All credits to @ashashik

### CODE REVIEW

**Comment by alignan** [comment]
Tested with  ~3.3V (meas 3.237) in ADC1

````
main(): This is RIOT! (Version: 2016.07-devel-669-g78209-vm-pull/boards/cc2538-adc)
RIOT ADC peripheral driver test
This test will sample all available ADC lines once every 100ms with
a 10-bit resolution and print the sampled results to STDIO

Successfully initialized ADC_LINE(0)
Successfully initialized ADC_LINE(1)
Successfully initialized ADC_LINE(2)
ADC_LINE(0): 3228
ADC_LINE(1): 233
ADC_LINE(2): 147
ADC_LINE(0): 3227
ADC_LINE(1): 231
ADC_LINE(2): 147
(...)
````

**Comment by alignan** [comment]
`tests/saul`:
````
main(): This is RIOT! (Version: 2016.07-devel-669-g78209-vm-ank-openmote-adc)
SAUL test application
Data:	[0] 3226
Data:	[0] 231
Data:	[0] 150
Data:	[0] 3224
Data:	[0] 232
Data:	[0] 148
Data:	[0] 3224
Data:	[0] 232
Data:	[0] 148
````

**Comment by alignan** [comment]
ping @kYc0o @haukepetersen

**Comment by OlegHahm** [comment]
@alignan, AFAIK @haukepetersen and @kYc0o are currently "unreachable" and only back at the end of the month. But maybe @PeterKietzmann or @A-Paul could take a look?

File: boards/cc2538dk/include/periph_conf.h [hunk 1]
```diff
@@ -151,6 +151,21 @@ static const periph_spi_conf_t spi_config[SPI_NUMOF] = {
 /** @} */
 
 /**
+ * @name ADC configuration
+ * @{
+ */
+#define ADC_NUMOF           (1)
+#define ADC_ALS_PIN         6
```
Comment (PeterKietzmann) [inline boards/cc2538dk/include/periph_conf.h]: This applies to all the periph_conf files that you adapted: Isn't there a way to be more consistent with the GPIO naming scheme? I don't find these names somehow speaking or at least self explaining. I won't know which pin to use without deeper research on that.

File: boards/remote-reva/include/periph_conf.h [hunk 2]
```diff
@@ -99,6 +99,24 @@ static const periph_spi_conf_t spi_config[SPI_NUMOF] = {
...
+ */
+#define ADC_NUMOF           (3)
+#define ADC1_PIN            5
+#define ADC2_PIN            4
+#define ADC3_PIN            2
+
+static const uint8_t periph_adc_map[ADC_NUMOF] = {
+  ADC1_PIN,
+  ADC2_PIN,
+  ADC3_PIN,
```
Comment (PeterKietzmann) [inline boards/remote-reva/include/periph_conf.h]: I just tested on a remote-reva but it seems ADC_3 does not work correctly or is at least calibrated completely wrong. Applying 0V level to this pin results in `95`, applying 3,3V results in `1200`. In contrast to that, the other channels sample `0` and `3276` as expected.

File: cpu/cc2538/include/periph_cpu.h [hunk 3]
```diff
@@ -73,6 +73,31 @@ typedef struct {
...
+/**
+ * @brief   Override resolution options
+ */
+#define HAVE_ADC_RES_T
+typedef enum {
+    ADC_RES_6BIT  = 0xa00,          /**< not supported by hardware */
+    ADC_RES_7BIT  = (0 << 4),       /**< ADC resolution: 7 bit */
+    ADC_RES_8BIT  = 0xb00,          /**< not supported by hardware */
+    ADC_RES_9BIT  = (1 << 4),       /**< ADC resolution: 9 bit */
+    ADC_RES_10BIT = (2 << 4),       /**< ADC resolution: 10 bit */
```
Comment (PeterKietzmann) [inline cpu/cc2538/include/periph_cpu.h]: Using `tests/periph_adc` the ADCs are initialized with a sampling width of 10 bit. However, applying 3,3 V to `ADC_2` runs into [this case](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_adc/main.c#L55). I didn't check if this is in cause of the normalisation to mV or the actual sample.

**Comment by PeterKietzmann** [comment]
Running `tests/periph_adc` it appears that the execution geopsts stuck/st. Didn't look behind it but the driver has "just" one blocking loop...
@alignan thanks for this PR and enhancing cc2538 support in RIOT. The code looks great and there are just few things to answer and/or address.

File: cpu/cc2538/periph/adc.c [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+    cc2538_soc_adc_t *adcaccess = SOC_ADC;
+    int16_t result;
+
+    /* Note - This has been hard coded .
+     *  Can choose from any of the choices below:
+     *  SOC_ADC_ADCCON_REF_INT or
+     *  SOC_ADC_ADCCON_REF_EXT_SINGLE or
+     *  SOC_ADC_ADCCON_REF_AVDD5
+     */
+    uint8_t refvoltage = SOC_ADC_ADCCON_REF_AVDD5;
```
Comment (alignan) [inline cpu/cc2538/periph/adc.c]: Yes, agree :-)

File: cpu/cc2538/periph/adc.c [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    while ((adcaccess->cc2538_adc_adccon1.ADCCON1 &
+        adcaccess->cc2538_adc_adccon1.ADCCON1bits.EOC) == 0);
+
+    /* Read conversion result, reading SOC_ADC_ADCH last to clear
+        SOC_ADC_ADCCON1.EOC */
+    result  = (((adcaccess->ADCL) & 0xfc));
+    result |= (((adcaccess->ADCH) & 0xff) << 8);
+
+    /* Return conversion result in mV */
+    return result / 10;
```
Comment (alignan) [inline cpu/cc2538/periph/adc.c]: OK

File: boards/cc2538dk/include/periph_conf.h (hunk 1 above)
Comment (alignan) [inline boards/cc2538dk/include/periph_conf.h]: I took as reference the CC2538DK documentation as default that pin is wired to the Ambient Light Sensor, but in practice it would also be required to power up the sensor via GPIO, but perhaps this should be done in another PR.  Do you have a proposal to name these pins?

File: boards/remote-reva/include/periph_conf.h (hunk 2 above)
Comment (alignan) [inline boards/remote-reva/include/periph_conf.h]: Yes, this pin has a 5/3 voltage divider, meant to connect 5VDC sensors... I think I commented this somewhere in the platform, but it should be more explicit here, I will add the comment, thanks!

**Comment by smlng** [comment]
ping, I'd like to see this one in the upcoming release + #5840 of course! @alignan, @PeterKietzmann what missing to merge this one, can we get it done (somehow soonish)?

**Comment by miri64** [comment]
Maintainers, please also update the cards when you update the milestone.

**Comment by miri64** [comment]
Postponed since Feature Freeze is closing in and there was no activity in this PR since the label was set.

**Comment by alignan** [comment]
Rebased and updated with comments addressed, for the next release milestone :smile:

**Comment by miri64** [comment]
(error is due to #6036).

**Comment by alignan** [comment]
Thanks @miri64 for the heads-up

**Comment by PeterKietzmann** [comment]
@alignan I will check this one again now. Please rebase

**Comment by PeterKietzmann** [comment]
Fortunately the driver doesn't block any more :-). But I still don't understand how the driver acts with different word sizes. I can't see a difference applying e.g. 3,3V to one pin using 12 Bit, 10 Bit, 8 Bit. In addition to that (even if I'm not yet sure this is reasonable), the [ADC API](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/adc.h#L131) does not allow negative sample values. However, I often land [here](https://github.com/RIOT-OS/RIOT/blob/master/tests/periph_adc/main.c#L55) when applying GND to the pins.

**Comment by alignan** [comment]
I'll double check and rebase

**Comment by alignan** [comment]
Mmm I have tried changing the resolution as `adc_res_t` yields no effect as @PeterKietzmann  commented earlier, but the configuration values seems OK.  I will need more time to test

File: cpu/cc2538/include/periph_cpu.h [hunk 6]
```diff
@@ -83,6 +83,31 @@ typedef struct {
...
+} adc_res_t;
+
+/**
+ * @brief Masks for getting data
+ * @{
+ */
+#define SOCADC_7_BIT_RSHIFT        9 /**< Mask for getting data( 7 bits ENOB) */
+#define SOCADC_9_BIT_RSHIFT        7 /**< Mask for getting data( 9 bits ENOB) */
+#define SOCADC_10_BIT_RSHIFT       6 /**< Mask for getting data(10 bits ENOB) */
+#define SOCADC_12_BIT_RSHIFT       4 /**< Mask for getting data(12 bits ENOB) */
```
Comment (jasonatran) [inline cpu/cc2538/include/periph_cpu.h]: Since there is a RSHIFT required after a conversion result, I suggest encoding the RSHIFT values into the ADC_RES_XXX values. For example:

ADC_RES_7BIT = (0 << 4) && (SOCADC_7_BIT_RSHIFT << 8);

After reading the conversion result, you can:

result = result >> (res >> 8);

**Comment by denzelwamburu** [comment]
What's the status with this?

**Comment by denzelwamburu** [comment]
Am trying to port the OS to cc1310 and cc1350

File: cpu/cc2538/include/periph_cpu.h (hunk 6 above)
Comment (alignan) [inline cpu/cc2538/include/periph_cpu.h]: Reasonable

**Comment by alignan** [comment]
The ADC conversion was working but somehow changing the resolution (i.e to 8-bits) was not properly being done (I checked the registers and everything seemed OK...).  You are free to use as you wish, port away.

**Comment by aabadie** [comment]
ping @alignan, this PR needs rebase. But let's try to have it in the next release !

**Comment by smlng** [comment]
closing, as #7327 was merged
